design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,217 @@
+
+Pros:
+* Cleaner user experience for the VM owner. No prior knowledge of CPUs added behind the scene is needed.
+* VM configuration is less complex.
+* As this is an alpha annotation, it can be removed/ignored when and if solutions inside/outside Kubevirt such as CPU Manager handles the situation differently.
+
+Cons:
+* Setting an annotation in kubevirt CR is unconventional. 
+* A cluster-wide configuration such as this may be CPU-wasteful for heterogeneous clusters, where node-level granularity is needed.
+
+# Fix Alternatives
+## Alternative #1 Adding `EmulatorThreadEvenParity` annotation per VMI
+Another way to do it is by adding the annotation on the VMI's annotations:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    metadata:
+      annotations:
+        kubevirt.io/EmulatorThreadEvenParity:
+    spec:
+      domain:
+        cpu:
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+          sockets: 1
+          cores: 4
+          threads: 2
+```
+Setting this will set the same behavior as on the suggested solution on virt-launcher and virt-handler.
+As opposed to the suggested solution where the user needs to add this annotation once on the kubevirt CR, it is possible to make this behavior VMI specific. This removes the assumption that the Nodes are homogeneous.
+The annotation will still be accompanied by the feature-gate configuration.
+
+Pros:
+* Matches the current crio convention of adding annotations in order to get high-performance pods, like `cpu-load-balancing.crio.io: disable`
+* Reduces the “wasted” CPUs to minimum possible per scenario, i.e. only for VMs that specifically add this annotation, and only when they ask for even CPUs.
+* As this is an alpha annotation, it can be removed/ignored when and if solutions inside/outside Kubevirt such as CPU Manager handles the situation differently.
+
+Cons:
+* We’re adding yet another semi alpha knob to an already complicated set of knobs.
+
+## Alternative #2 Configure both pod and Guest
+It is possible to use both the current Guest API and the old yet existing pod API, like so:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        resources:
+          requests:
+            cpu: '10'
+          limits:
+            cpu: '10'
+        cpu:
+          cores: 4
+          sockets: 1
+          threads: 2
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+```
+In order to do that the current validating webhooks regarding dedicatedCpuPlacement and IsolatedEmulatorThread will have to be loosened.
+
+Pros:
+* This option does not introduce any alpha knobs.
+
+Cons:
+* This option exposes the user to unnecessary implementation details and possible waste of physical CPUs. They have to do things manually, opening the door for misuse. Loosening the validating webhook doesn't make sense in the general case.
```

Comment:
If we did not have the webhook, we would not have needed this proposal at all. VM owners that want a specific size for their virt-launcher could have done so themselves, just like that can already do for VMs without isolatedEmulatorThreads.

None of us knows how much misuse this webhook prevented, but we do know how much effort it consts us now. IMO this is prime example where over-protection is harmful.

Reply:
I think the intention of the original feature developers was to "make sure no one is using it in a way that it was not intended", thus opening a way for bugs that come from miss use. 

However, I do see your point here. Perhaps we could all benefit from sharing this example to the team so that we could converse and align with this for future features.